### ISSUE TITLE
Association variables not available in resource link callbacks

### ISSUE THREAD

**Issue by john-griffin** [JSONAPI-Resources/jsonapi-resources#184]
I have been trying to build some authorization into a resource using before callbacks but have been struggling to do so with linked/associated resources due to lack of available association variables. When any of these callbacks are fired:

``` ruby
:create_has_many_link
:replace_has_many_links
:create_has_one_link
:replace_has_one_link
:remove_has_many_link
:remove_has_one_link
```

it does not seem any variables for the link/association can be accessed.

For example, say I was trying to associate a `Comment` with a `User` on a `CommentResource`. In the `before` `create_has_one_link` callbacks there are no available variables that include which `User` is being set. But I can access the `Comment` on the `@model`.

Is this the expected behaviour or am I missing something?

**Comment by lgebhardt** [comment]
It is expected behavior. I'm not sure how to pass variables into the callbacks (we're using ActiveSupport::Callbacks). I'll take a look into whether it's something that's easy and would make sense in all cases.

You could use the callbacks on the OperationsProcessor and intercept the `@opperation` variable.

**Comment by john-griffin** [comment]
Thanks. I started overriding attr assignment methods in the resource to perform the authorization. @brettchalupa has an interesting solution in his PR.

What is the best way to build callbacks on `OperationsProcessor`? I can't find a natural point in our app to add them.

**Comment by lgebhardt** [comment]
I'm looking at @brettchalupa's solution now. Using the `around_replace_fields` callback might give you what you want.

To add callbacks on the `OperationsProcessor` you will need to create a custom `OperationsProcessor` derived from the `JSONAPI::ActiveRecordOperationsProcessor` (or the base one if not using AR). You will then need to override the `create_operations_processor` method on your controller:

``` ruby
    def create_operations_processor
#      JSONAPI::ActiveRecordOperationsProcessor.new
      CustomActiveRecordOperationsProcessor.new
    end
```

We should get that into the docs. Sorry for the omission.

**Comment by john-griffin** [comment]
No worries, thanks for the details.

**Comment by lgebhardt** [comment]
@john-griffin This was closed when I merged #187. Feel free to reopen if needed.
